lib/config/dynamic/admiral/source.go

```diff
@@ -0,0 +1,396 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package admiral
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/go-openapi/runtime"
+	rtclient "github.com/go-openapi/runtime/client"
+	strfmt "github.com/go-openapi/strfmt"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+
+	"github.com/vmware/govmomi/vim25/types"
+	vchcfg "github.com/vmware/vic/lib/config"
+	"github.com/vmware/vic/lib/config/dynamic"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client/config_registries"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client/projects"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client/resources_compute"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/models"
+	"github.com/vmware/vic/pkg/vsphere/session"
+	"github.com/vmware/vic/pkg/vsphere/tags"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+const (
+	VicProductCategory = "VsphereIntegratedContainers"
+	ProductVMTag       = "ProductVM"
+	admiralEndpointKey = "guestinfo.vicova.admiral.endpoint"
+
+	clusterFilter = "(address eq '%s' and customProperties.__containerHostType eq 'VCH')"
+)
+
+// #nosec
+const admiralTokenKey = "guestinfo.vicova.engine.token"
+
+var (
+	trueStr        = "true"
+	projectsFilter = "customProperties.__enableContentTrust eq 'true'"
+)
+
+// NewSource creates a new Admiral dynamic config source. sess
+// is a valid vsphere session object. vchID is a unique identifier
+// that will be used to lookup the VCH in the Admiral instance; currently
+// this is a URI to the docker endpoint in the VCH.
```

Comment:
Yes, that is why the endpoint URL is not cached at all in the persona; DHCP changes to the IP address should get picked up since `client.localhost` does get updated when the IP changes.  Of course Admiral doesn't update the VCH when its IP changes, and yes, we need to sit down with them to figure out a better way.